californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java

```diff
@@ -874,12 +889,15 @@ private void handleBlock2Response(final Exchange exchange, final Response respon
 			exchange.getRequest().cancel();
 
 		} else {
-
-			Block2BlockwiseStatus status = getInboundBlock2Status(key, exchange, response);
-
-			if (block2.getNum() == status.getCurrentNum() && (block2.getNum() == 0 || response.getToken().equals(exchange.getCurrentRequest().getToken()))) {
-
-				// check token to avoid mixed blockwise transfers (possible with observe) 
+			Block2BlockwiseStatus status;
+			synchronized (block2Transfers) {
```

Comment:
That ensure, the the status is either the right one, cleared and renewed, or the response is ignored. it must be executed in a snyc block.

Reply:
There is one thing that I don't understand : 
 it seems we already checked before  `if (discarding ..` in `receiveResponse`. :thinking:

Reply:
Yes, but between the first and the second check, a new notification may change the result :-). I still try to check, if a final "!status complete" check, before sending the next request is a further improvement ...

Reply:
Just to mention:
That seems to be race condition.